scalalib/src/mill/scalalib/CoverageModule.scala

```diff
@@ -0,0 +1,309 @@
+package mill.scalalib
+
+import scoverage.{Constants => ScoverageConstants};
+
+import mill.Cross
+import mill.scalalib.{SbtModule, PublishModule, Dep, CrossSbtModule, DepSyntax}
+import mill.scalalib.publish.{PomSettings, License, Developer, SCM}
+import mill.define.TaskModule
+import scala.sys.process.Process
+import mill._, scalalib._
+import mill.util.Logger
+import ammonite.ops._
+import java.io.File
+
+import scoverage.IOUtils;
+import scoverage.Serializer;
+import scoverage.Coverage;
+import scoverage.report.{
+  CoverageAggregator,
+  CoberturaXmlWriter,
+  ScoverageHtmlWriter,
+  ScoverageXmlWriter
+}
+
+/** A little trait to mix in within a module to add coverage.
+  *
+  * {{{
+  * object coverage extends Tests with CoverageModule {
+  *     def testIvyDeps = Agg(ivy"org.scalatest::scalatest:3.0.1")
+  *     def testFrameworks = Seq("org.scalatest.tools.Framework")
+  * }
+  * }}}
+  *
+  * */
+trait CoverageModule extends ScalaModule with CoverageOptions {
+  import CoverageModule._
+  import CoverageUtil._
+
+  final override def ivyDeps =
+    super.ivyDeps() ++ Agg(ivyScoverageRuntime) ++ testIvyDeps()
+
+  def testIvyDeps = T { Agg.empty[Dep] }
+
+  override def compileIvyDeps = super.compileIvyDeps() ++ Agg(ivyScoverageRuntime)
+
+  override def scalacPluginIvyDeps =
+    super.scalacPluginIvyDeps() ++ Agg(ivyScoveragePlugin)
+
+  override def scalacOptions = T {
+    val pluginPath = {
+      val cp = scalacPluginClasspath()
+      cp.find { pathRef =>
+        val fileName = pathRef.path.toNIO.getFileName.toString
+        fileName.toString.contains(ScalacPluginArtifact) && fileName.toString
+          .contains(scoverageVersion)
+      }
+        .getOrElse {
+          throw new Exception(
+            s"Fatal: ${ScalacPluginArtifact} (${scoverageVersion}) not found on the classpath:\n\t" + cp
+              .map(_.path)
+              .mkString("\n\t"))
+        }
+    }.path
+    super.scalacOptions() ++ Seq(
+      Some(s"-Xplugin:${pluginPath}"),
+      Some(s"-P:scoverage:dataDir:${coverageDataDir().toNIO}"),
+      Option(excludedPackages).map(v => s"-P:scoverage:excludedPackages:$v"),
+      Option(excludedFiles).map(v => s"-P:scoverage:excludedFiles:$v"),
+      if (highlighting) Some("-Yrangepos") else None // rangepos is broken in some releases of scala so option to turn it off
+    ).flatten
+  }
+
+  private def coverageRootDir = T {
+    // FIXME: using artifactName is wrong
+    val path = millSourcePath / 'out / artifactName()
+      .split('-')
+      .last / scalaVersion() / 'coverage / 'dest
```

Comment:
Finding the destination path for the `coverage` target (or whatever the user will call it) would be nice.